cdm/core/src/main/java/ucar/nc2/internal/iosp/hdf5/H5headerNew.java

```diff
@@ -1553,19 +1547,73 @@ private boolean makeVariableShapeAndType(Group.Builder parent, Variable.Builder
-        parent.addEnumTypedef(enumTypedef);
+      // dmh: An HDF5 file, at least as used by netcdf-4, may define an enumeration
+      // type one or more times:
+      // 1. There may be an explicit, independent enum type definition.
+      // 2. A variable/HDF5-Dataset may define an implicit enum type with the same name as the variable.
+      // 3. A variable may define an implicit enum type that is a copy of a case 1 enum type;
+      // the implicit enum type will have the same name as the independent enum type.
+      //
+      // The algorithm to infer (and if necessary, create) the proper EnumTypeDef is as follows:
+      // Step 1. If there exists a case 1 enum type with the same name as the variable's enum type,
+      // then use that.
+      // Step 2. If the variable's enum type has the same name as the variable, then we need to
+      // look for a case 1 enum type that is structurally the same as the variable's enum type.
+      // If such exists, then use that.
+      // Step 3: Otherwise, create a new enum type and use that. The new enum type
+      // will have these properties:
+      // a. It is defined in the same group as the variable
+      // b. It has a mutated name similar to the variable's name, namely <variablename>_enum_t.
+
+      EnumTypedef actualEnumTypedef = null; // The final chosen EnumTypedef
+
+      // Step 1:
+      // See if an independent enum type already exists with the same name
+      Optional<EnumTypedef> candidate = parent.findEnumTypedef(mdt.enumTypeName, true);
+      if (candidate.isPresent()) {
+        // There is an independent type, so use it.
+        actualEnumTypedef = candidate.get();
+      }
+
+      // Step 2:
+      // See if an independent enum type already exists that is structurally similar.
+      if (actualEnumTypedef == null && mdt.enumTypeName.equals(v.shortName)) {
+        // Materialize a enum type def for search purposes; name is irrelevant
+        EnumTypedef template = new EnumTypedef(mdt.enumTypeName, mdt.map);
+        // Search for a structurally similar enum type def
+        candidate = parent.findSimilarEnumTypedef(template, true);
+        if (candidate.isPresent()) {
+          // There is an independent type, so use it.
+          actualEnumTypedef = candidate.get();
+        }
       }
-      v.setEnumTypeName(enumTypedef.getShortName());
-    }
 
+      // Step 3: Create an independent type
+      if (actualEnumTypedef == null) {
+        String newname = null;
+        if (mdt.enumTypeName.equals(v.shortName)) {
+          // Create mutated name to avoid name conflict
+          newname = mdt.enumTypeName + "_enum_t";
+        } else {
+          newname = mdt.enumTypeName;
+        }
+        actualEnumTypedef = new EnumTypedef(newname, mdt.map);
+        // Add to the current group(builder)
+        parent.addEnumTypedef(actualEnumTypedef);
+      }
+
+      if (actualEnumTypedef == null) {
```

Comment:
I don't think this `if` can ever be true since above if `actualEnumTypedef == null` then it gets constructed

Reply:
done